graph/telemetry/istio/appender/aggregate_node.go

```diff
@@ -0,0 +1,276 @@
+)
+
+// AggregateNodeAppender is responsible for injecting aggregate nodes into the graph to gain
+// visibility into traffic aggregations for a user-specfied metric attribute.
+type AggregateNodeAppender struct {
+	Aggregate          string
+	AggregateValue     string
+	GraphType          string
+	InjectServiceNodes bool
+	Namespaces         map[string]graph.NamespaceInfo
+	QueryTime          int64 // unix time in seconds
+}
+
+// Name implements Appender
+func (a AggregateNodeAppender) Name() string {
+	return AggregateNodeAppenderName
+}
+
+// AppendGraph implements Appender
+func (a AggregateNodeAppender) AppendGraph(trafficMap graph.TrafficMap, globalInfo *graph.AppenderGlobalInfo, namespaceInfo *graph.AppenderNamespaceInfo) {
+	if len(trafficMap) == 0 {
+		return
+	}
+
+	if globalInfo.PromClient == nil {
+		var err error
+		globalInfo.PromClient, err = prometheus.NewClient()
+		graph.CheckError(err)
+	}
+
+	if a.AggregateValue == "" {
+		a.appendGraph(trafficMap, namespaceInfo.Namespace, globalInfo.PromClient)
+	} else {
+		a.appendNodeGraph(trafficMap, namespaceInfo.Namespace, globalInfo.PromClient)
+	}
+}
+
+func (a AggregateNodeAppender) appendGraph(trafficMap graph.TrafficMap, namespace string, client *prometheus.Client) {
+	log.Tracef("Resolving request aggregates for namespace=[%s], aggregate=[%s]", namespace, a.Aggregate)
+	duration := a.Namespaces[namespace].Duration
+
+	// query prometheus for aggregate info in two queries (assume aggregation is typically request classification, so use dest telemetry):
+	//   note1: we want to only match the aggregate when it is set and not "unknown".  But in Prometheus a negative test on an unset label
+	//      matches everything, so using %s!=unknown mneans we still have to filter out unset time-series below...
+	//   note2: for now we will filter out aggregates with no traffic on the assumption that users probably don't want to
+	//      see them and it will just increase the graph density.  To change that behavior remove the "> 0" conditions.
+	// 1) query for requests originating from a workload outside the namespace.
+	groupBy := fmt.Sprintf("source_workload_namespace,source_workload,source_%s,source_%s,destination_service_namespace,destination_service,destination_service_name,destination_workload_namespace,destination_workload,destination_%s,destination_%s,request_protocol,response_code,grpc_response_status,response_flags,%s", appLabel, verLabel, appLabel, verLabel, a.Aggregate)
+	httpQuery := fmt.Sprintf(`sum(rate(%s{reporter="destination",source_workload_namespace!="%s",destination_service_namespace="%v",%s!="unknown"}[%vs])) by (%s) > 0`,
+		"istio_requests_total",
+		namespace,
+		namespace,
+		a.Aggregate,
+		int(duration.Seconds()), // range duration for the query
+		groupBy)
+	tcpQuery := fmt.Sprintf(`sum(rate(%s{reporter="destination",source_workload_namespace!="%s",destination_service_namespace="%v",%s!="unknown"}[%vs])) by (%s) > 0`,
+		"istio_tcp_sent_bytes_total",
```

Comment:
I still don't understand this use case. And you are doing two more queries (to support istio_tcp_sent_bytes_total) so we need to make sure these really are going to be useful. Are you sure users can categorize tcp traffic in envoy filters? Is that a valid thing to do?

The only thing I can think of is for gRPC traffic -- is gRPC telemetry in the istio_tcp_sent_bytes_total timeseries? If so, how does one create envoy filters to categorize gRPC traffic (is that even possible)?

If it isn't possible or doesn't make sense, I would remove the code involving istio_tcp_sent_bytes_total since it just adds 2 more prom queries that are being executed but are not useful, and clutters up the code.

Reply:
After some itsio slack conversation it seems the gRPC may be less useful, or not useful, but it's a no-op for us because it's reported as a request counter in requests_total.  It seems the TCP support may not be needed, so I'll remove it because it does cost us some queries.  If someone shows it's useful we can always add it back (I may leave the queries commented out).

Reply:
PR Updated.